File: benchmarks/fixtures/webapp_go/internal/database/pool.go
package database

import (
    "fmt"
    "sync"

    "webapp_go/pkg/logger"
)

var poolLog = logger.GetLogger("database.pool")

// ConnectionHandle wraps a database connection with metadata.
type ConnectionHandle struct {
    ID       int
    InUse    bool
    Database string
}

// ConnectionPool manages a pool of database connections.
type ConnectionPool struct {
    connections []*ConnectionHandle
    maxSize     int
    mu          sync.Mutex
}

// NewConnectionPool creates a pool with the specified max size.
func NewConnectionPool(maxSize int) *ConnectionPool {
    poolLog.Info("Creating connection pool with max size: %d", maxSize)
    pool := &ConnectionPool{
        connections: make([]*ConnectionHandle, 0, maxSize),
        maxSize:     maxSize,
    }
    for i := 0; i < maxSize; i++ {
        pool.connections = append(pool.connections, &ConnectionHandle{
            ID:       i,
            InUse:    false,
            Database: "default",
        })
    }
    poolLog.Info("Connection pool initialized with %d connections", maxSize)
    return pool
}

// GetConnection acquires a connection from the pool.
func (p *ConnectionPool) GetConnection() (*ConnectionHandle, error) {
    p.mu.Lock()
    defer p.mu.Unlock()
    poolLog.Debug("Requesting connection from pool")
    for _, conn := range p.connections {
        if !conn.InUse {
            conn.InUse = true
            poolLog.Info("Acquired connection #%d", conn.ID)
            return conn, nil
        }
    }
    poolLog.Error("No available connections in pool")
    return nil, fmt.Errorf("connection pool exhausted")
}

// ReleaseConnection returns a connection to the pool.
func (p *ConnectionPool) ReleaseConnection(handle *ConnectionHandle) {
    p.mu.Lock()
    defer p.mu.Unlock()
    poolLog.Debug("Releasing connection #%d", handle.ID)
    handle.InUse = false
}

// ActiveCount returns the number of connections currently in use.
func (p *ConnectionPool) ActiveCount() int {
    p.mu.Lock()
    defer p.mu.Unlock()
    count := 0
    for _, conn := range p.connections {
        if conn.InUse {
            count++
        }
    }
    poolLog.Debug("Active connections: %d", count)
    return count
}

// Shutdown closes all connections in the pool.
func (p *ConnectionPool) Shutdown() {
    p.mu.Lock()
    defer p.mu.Unlock()
    poolLog.Info("Shutting down connection pool")
    for _, conn := range p.connections {
        conn.InUse = false
    }
    p.connections = p.connections[:0]
    poolLog.Info("Connection pool shut down")
}

File: benchmarks/fixtures/webapp_rs/database/pool.rs
use crate::utils::helpers::get_logger;

/// A handle representing a single borrowed connection from the pool.
pub struct ConnectionHandle {
    /// The connection identifier.
    pub id: String,
    /// Whether this handle is currently in use.
    pub in_use: bool,
    /// Number of queries executed on this handle.
    pub query_count: u64,
}

/// A pool of database connections with configurable size.
pub struct ConnectionPool {
    /// The data source name / connection string.
    dsn: String,
    /// Maximum number of connections in the pool.
    max_size: usize,
    /// All connection handles managed by this pool.
    connections: Vec<ConnectionHandle>,
    /// Whether the pool has been initialized.
    initialized: bool,
}

impl ConnectionPool {
    /// Create a new connection pool with the given DSN and size.
    pub fn new(dsn: &str, max_size: usize) -> Self {
        let logger = get_logger("database.pool");
        logger.info(&format!("Creating pool: dsn={}, size={}", dsn, max_size));
        Self {
            dsn: dsn.to_string(),
            max_size,
            connections: Vec::new(),
            initialized: false,
        }
    }

    /// Initialize the pool by creating all connections.
    pub fn initialize(&mut self) -> Result<(), String> {
        let logger = get_logger("database.pool");
        if self.initialized {
            return Ok(());
        }
        for i in 0..self.max_size {
            self.connections.push(ConnectionHandle {
                id: format!("conn-{}", i),
                in_use: false,
                query_count: 0,
            });
        }
        self.initialized = true;
        logger.info(&format!("Pool initialized with {} connections", self.max_size));
        Ok(())
    }

    /// Acquire a connection from the pool.
    pub fn get_connection(&mut self) -> Result<&mut ConnectionHandle, String> {
        let logger = get_logger("database.pool");
        if !self.initialized {
            self.initialize()?;
        }
        for conn in &mut self.connections {
            if !conn.in_use {
                conn.in_use = true;
                conn.query_count += 1;
                logger.info(&format!("Acquired connection {}", conn.id));
                return Ok(conn);
            }
        }
        Err("Connection pool exhausted".to_string())
    }

    /// Release a connection back to the pool by ID.
    pub fn release_connection(&mut self, conn_id: &str) {
        let logger = get_logger("database.pool");
        for conn in &mut self.connections {
            if conn.id == conn_id {
                conn.in_use = false;
                logger.info(&format!("Released connection {}", conn_id));
                return;
            }
        }
    }

    /// Return pool statistics.
    pub fn stats(&self) -> (usize, usize, usize) {
        let active = self.connections.iter().filter(|c| c.in_use).count();
        let idle = self.connections.len() - active;
        (self.connections.len(), active, idle)
    }

    /// Shut down the pool and drop all connections.
    pub fn shutdown(&mut self) {
        let logger = get_logger("database.pool");
        self.connections.clear();
        self.initialized = false;
        logger.info("Pool shut down");
    }
}
